src/index/writer.rs

```diff
@@ -0,0 +1,235 @@
+
+    #[snafu(display("Condition label is not valid due to: {}", msg))]
+    InvalidCondition { msg: String },
+
+    #[snafu(display("Internal error : {}", msg))]
+    Internal { msg: String },
+}
+
+/// Index Writer writes a fixed size Record to the stream.
+///
+/// Write takes a byte array as data and a Label. It hashes the Label entry key and construct a Record. Then
+/// it serializes the Record and writes to the stream.
+///
+/// # Examples
+/// ```no_run
+/// use pravega_client_config::ClientConfigBuilder;
+/// use pravega_client::client_factory::ClientFactory;
+/// use pravega_client_shared::ScopedSegment;
+/// use pravega_client_macros::Label;
+/// use std::io::Write;
+/// use tokio;
+///
+/// #[derive(Label, Debug, PartialOrd, PartialEq)]
+/// struct MyLabel {
+///     id: u64,
+///     timestamp: u64,
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     // assuming Pravega controller is running at endpoint `localhost:9090`
+///     let config = ClientConfigBuilder::default()
+///         .controller_uri("localhost:9090")
+///         .build()
+///         .expect("creating config");
+///
+///     let client_factory = ClientFactory::new(config);
+///
+///     // assuming scope:myscope, stream:mystream and segment 0 do exist.
+///     let segment = ScopedSegment::from("myscope/mystream/0");
+///
+///     let mut index_writer = client_factory.create_index_writer(segment).await;
+///
+///     let label = MyLabel{id: 1, timestamp: 1000};
+///     let data = vec!{1; 10};
+///
+///     index_writer.append(data, label).await.expect("append data with label");
+///     index_writer.flush().await.expect("flush");
+/// }
+/// ```
+pub struct IndexWriter<T: Label + PartialOrd + PartialEq + Debug> {
+    byte_writer: ByteWriter,
+    entries: Option<Vec<(u128, u64)>>,
+    label: Option<T>,
+    _label_type: PhantomData<T>,
+}
+
+impl<T: Label + PartialOrd + PartialEq + Debug> IndexWriter<T> {
+    pub(crate) async fn new(factory: ClientFactory, segment: ScopedSegment) -> Self {
+        let mut byte_writer = factory.create_byte_writer_async(segment.clone()).await;
+        byte_writer.seek_to_tail_async().await;
+
+        let index_reader = factory.create_index_reader(segment.clone()).await;
+        let tail_offset = index_reader.tail_offset().await.expect("get tail offset");
+        let head_offset = index_reader
+            .head_offset()
+            .await
+            .expect("get readable head offset");
+        if head_offset != tail_offset {
+            let prev_record_offset = tail_offset - RECORD_SIZE;
+            let record = index_reader
+                .read_record_from_random_offset(prev_record_offset)
+                .await
+                .expect("read last record");
+
+            return IndexWriter {
+                byte_writer,
+                entries: Some(record.entries),
+                label: None,
+                _label_type: PhantomData,
+            };
+        }
+        IndexWriter {
```

Comment:
This is the same inside and outside the `if`. Maybe just has a local variable for `entries`?